### ISSUE TITLE
Performance Optimizations

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/py-evm#542]
### What is wrong?

There is rudamentary full sync code in place and with that, we're finding performance issues.

- https://drive.google.com/open?id=1OwLcvS-ONGmgucL4Tccf5nqP2r_QCglv
- [github-mirror](https://github.com/ethereum/py-evm/files/1895559/block-import-stats.txt)

The link above is the `cProfile` output which can be loaded using the [`pstats`](https://docs.python.org/3.5/library/profile.html) standard library module to see where time is spent.

### Known "low hanging fruit"

The following things are where we're seeing the *most* time spent which can likely be optimized.

- Remove support for python 2 from `pyrlp`
- Further `pyrlp` optimizations
- Caching of the `VMState.prev_hashes` generation.

### Ways to make things faster

[Cython](http://cython.org/) looks to be a very promising way for us to do certain optimizations.  We should spend a little time doing *discovery* to see how this well this works and what kind of performance gains we can get.

**Comment by gsalgado** [comment]
Another thing I've just noticed is that after applying a transaction we seem to throw away objects representing the state and then re-read (and decode) them from the DB in order to apply the next block's tx. One example is the block's receipts: ISTM that for every tx we append the generated receipts to `BaseState.receipts`, but then for the next transaction we re-read all those receipts from the DB. Would it be possible to preserve those `Receipt` instances (and maybe other things?) across transaction boundaries?

**Comment by pipermerriam** [comment]
> Would it be possible to preserve those Receipt instances (and maybe other things?) across transaction boundaries?

Yes, though I'd like to try and accomplish this with caching as I like the way we currently have things with minimal statefullness.

**Comment by gsalgado** [comment]
So, with all the changes we made to py-evm/py-rlp/py-trie things got significantly better, but `JournalDB` is still the biggest offender and makes block importing just plain unfeasible (because it does an extra `LevelDB.get()` every time we update/delete a key) on a fully-synced ropsten DB. Here's a 
[new stats file](https://github.com/ethereum/py-evm/files/1908308/block-import-stats.txt) that aggregates the profiling stats of only `Chain.import_block()` for 8 consecutive blocks (unlike the previous one, which was profiling the networking code as well).

Given the above, I've been thinking about ways to fix that and I believe we might be able to use [LevelDB snapshots](http://plyvel.readthedocs.io/en/latest/api.html#Snapshot) in our `JournalDB` to avoid the extra `LevelDB.get()` on every key update/delete. We'd change `JournalDB` to add only the keys to the journal, and then on `.revert()` we read their values from the snapshot and set those values to the actual DB. That way there's no overhead on DB operations unless we have to revert to a snapshot.

Can you guys see foresee any issues with that approach? Do you think it's worth pursuing?

**Comment by pipermerriam** [comment]
@gsalgado my only *worry* is related to exposing underlying database architecture to higher level APIs since currently, we assume our database is a *dumb* key/value store.

In general though, I have no opposition to this.

**Comment by pipermerriam** [comment]
Posted this in gitter.  Leaving it here too:

I'm wondering if we could accomplish a similar journal mechanism using a combination of the [`BatchDB`](https://github.com/ethereum/py-evm/issues/160) concept, and just rolling back the `state_root` on reversions instead of tracking all the actual database changes.

- set a checkpoint in the batch db, take note of the `state_root`
- revert means throwing away the batched (but uncommitted) changes going back to the checkpoint.
- commit means committing the batch.

**Comment by gsalgado** [comment]
> my only worry is related to exposing underlying database architecture to higher level APIs since currently, we assume our database is a dumb key/value store.

Well, it would all still be kept confined within `JournalDB` so not really exposed anywhere, right? And we could have multiple `JournalDB` implementations, one for snapshot-able backends and another for dumb key/value stores.

**Comment by gsalgado** [comment]
So, I've implemented our [JournalDB using LevelDB snapshots](https://github.com/gsalgado/py-evm/commit/badbe15121189b871c2422a003c13f5121b8081a) but for **some blocks** we seem to be doing a lot of `JournalDB.revert()`s, so although we no longer make a `LevelDB.get()` call every time we set/delete a key, when we do a `JournalDB.revert()` we make tons of `Snapshot.get()` calls so there's no performance gains when importing **those blocks**.

Using BatchDB as suggested by @pipermerriam may yield better performance improvements as, IIUC, in that case a revert is pretty much free

**Comment by gsalgado** [comment]
Other changes to enhance performance: #565 #573, #569

**Comment by carver** [comment]
According to [these benchmarks](http://www.lmdb.tech/bench/microbench/benchmark.html), we can probably get some marginal improvements from bumping up from the default cache size and using [batch writes](http://plyvel.readthedocs.io/en/latest/api.html#write-batch).

Batch writes in particular could speed up write operations by 35% from 164k ops/s to 221k ops/s.

**Comment by pipermerriam** [comment]
@carver that suggests a new underlying base db API for submitting batch writes to the database.  Default implementation can just be to iterate over them and do singular writes, but on our leveldb backends we can leverage the batching functionality (as well as with our journal db cc @cburgdorf )

**Comment by pipermerriam** [comment]
Also, that suggests that my thoughts on using sqlite could have a noticeable hit on performance.

**Comment by carver** [comment]
> that suggests a new underlying base db API for submitting batch writes to the database ... as well as with our journal db

Yeah, I think we can keep the same `db.set()` api on the in-memory DB, and then do a single batch-write at `JournalDB` commit time when exiting the context.

**Comment by pipermerriam** [comment]
We may be talking about the same thing.

https://github.com/ethereum/py-evm/blob/master/evm/db/backends/base.py

I think this needs a new API:  `BaseDB.batch_set(...)`

`MemoryDB` would implement this as a loop calling down to individual `set` calls.  The LevelDB backend would use the batch APIs from plyvel.

**Comment by gsalgado** [comment]
Fresh cProfile stats, now with all performance improvements and a much shorter DB (thanks to #600): [block-import-stats.txt](https://github.com/ethereum/py-evm/files/1955545/block-import-stats.txt)

TLDR: trie seems to be our bottleneck now, together with some `ChainDB` methods, but some of the latter can probably be cached
